Implement the session lock protocol

lcolitti opened the pull request:
Implement the session lock protocol

Currently implemented:

- Locking and unlocking.
- Display above all desktop layers, including OVERLAY (but below
  DWIDGET).
- Waiting for the client to display locked surfaces before
  declaring the screen to be locked. If the client does not
  display all surfaces, locks anyway after a timeout.
- Multiple outputs, including outputs being added/removed while
  locking or locked.
- Leaving the screen locked if the client crashes (but allows a
  new client to come in and lock the screen.

TODO:
- Deal with outputs being resized.
- Properly display text if the screenlocker crashes. Currently it
  just displays two a big "explosion" emoji.
- Use a view instead of a plain wlr_surface_node?

Tested with swaylock.

Fixes #1494
Fixes #1358

soreau commented on plugins/protocols/session-lock.cpp:
```diff
@@ -0,0 +1,198 @@
+                auto layer_node = wo->node_for_layer(target_layer);
+
+                (*node)->set_enabled(true);
+                wf::scene::add_front(layer_node, *node);
```
You might need to set this to `add_back` and/or change the `target_layer` above, to get swaylock to appear.

lcolitti replied:
Changed to add_back. Not sure what the difference between front and back would be here. I guess it doesn't matter because the only thing in the lock layer is this surface?

ammen99 commented on src/api/wayfire/core.hpp:
```diff
@@ -144,6 +144,8 @@ class compositor_core_t : public wf::object_base_t, public signal::provider_t
         wlr_xdg_foreign_v1 *foreign_v1;
         wlr_xdg_foreign_v2 *foreign_v2;
+
+        wlr_session_lock_manager_v1 *session_lock;
```
Do we need to define this in core, since the functionality is implemented as a plugin?

lcolitti replied:
Left over from a previous attempt :-). Removed.

ammen99 commented on plugins/protocols/session-lock.cpp:
```diff
@@ -0,0 +1,198 @@
+
+            void handle_keyboard_enter(wf::seat_t *seat)
+            {
+                // TODO: use a keyboard grab instead?
```
I think this is fine - wlroots grabs receive events if Wayfire decides to forward requests to wlroots - with HIGH focus importance and putting the lockscreen above everything else, Wayfire should always feed the focus to the lockscreen first.

lcolitti replied:
BTW, any reason not to use wf::input_manager_t::set_exclusive_focus here? Seems like it would be simpler. If the client crashes I assume that focus will go nowhere, but it sounds like that's what we want.

ammen99 replied:
set_exclusive_focus was meant for the older implementation of screenlockers, I think we can even remove this function nowadays. It is better to use the scenegraph API for this ;)

ammen99 commented on plugins/protocols/session-lock.cpp:
```diff
@@ -0,0 +1,198 @@
+        }
+
+      private:
+        wf::keyboard_interaction_t *interaction;
```
I think this should be a unique_ptr

ammen99 commented on plugins/protocols/session-lock.cpp:
```diff
@@ -0,0 +1,198 @@
+                uint32_t serial = wlr_session_lock_surface_v1_configure(lock_surface,
+                    lock_surface->output->width,
+                    lock_surface->output->height);
+                LOGI("surface_configure output=", lock_surface->output->name, " ",
```
Once we're ready to merge, this should be LOGD or maybe even LOGC(VIEWS, ...)

ammen99 commented on plugins/protocols/session-lock.cpp:
```diff
@@ -0,0 +1,198 @@
+                auto wo = wf::get_core().output_layout->find_output(lock_surface->output);
+                auto layer_node = wo->node_for_layer(target_layer);
+
+                (*node)->set_enabled(true);
```
Is this necessary? I thought nodes are created enabled by default ..

lcolitti replied:
Removed.

ammen99 commented on plugins/protocols/session-lock.cpp:
```diff
@@ -0,0 +1,198 @@
+            {
+                LOGI("new_lock: already locked");
+                wlr_session_lock_v1_destroy(wlr_lock);
+            }
```
The old locking implementation also inhibits the output so that no plugins can be activated (I think this implementation doesn't do that?): https://github.com/WayfireWM/wayfire/blob/2e0926f8f5a1c3834c51e97673bfbf074f4506ac/src/output/output-impl.hpp#L67-L79

We'd probably want to move these functions to the public API so that lockscreens can be implemented from plugins as well. Also, please change inhibit/uninhibit to `set_inhibited(bool)` in the process :)

lcolitti replied:
Done in [this commit](https://github.com/WayfireWM/wayfire/pull/2162/commits/714cd91a88758cca1237fc2388ef0e31e90fe4d4). Let me know if you'd like me to split out these refactoring commits to another PR so we can merge them first.

ammen99 replied:
> Let me know if you'd like me to split out these refactoring commits to another PR so we can merge them first.

We can do it as you wish. For my personal PRs I usually use git rebase and squash related commits, but if you find multiple PRs simpler, we can do that too.

ammen99 commented on plugins/protocols/session-lock.cpp:
```diff
@@ -0,0 +1,198 @@
+
+                    lock_surface_destroy.disconnect();
+                });
+                lock_surface_destroy.connect(&lock_surface->events.destroy);
```
You will need one listener per created surface (because listeners cannot connect to multiple object's signals unlike Wayfire's signals). So lock_surface_destroy() probably will have to be moved to the node class.

lcolitti replied:
Thanks for the suggestion. Done, I think.

ammen99 commented on plugins/protocols/session-lock.cpp:
```diff
@@ -0,0 +1,198 @@
+                new_surface.disconnect();
+                unlock.disconnect();
+                destroy.disconnect();
+                LOGI("destroy");
```
I think that the point of ext-screenlock was that we should keep the screen locked if the client crashes and we haven't unlocked.

ammen99 commented on plugins/protocols/session-lock.cpp:
```diff
@@ -0,0 +1,198 @@
+            if (cur_lock.get() == nullptr)
+            {
+                cur_lock.reset(new wayfire_session_lock(this, wlr_lock));
+                // TODO: blank the screen, or set a timeout for the lock app to commit a surface.
```
Maybe same as in the 'crashed' case, I would create a simple cairo surface saying 'Waiting for screenlock' or 'Screenlocker crashed?' 

Here is a small example of how to show a cairo-rendered text in a node: https://github.com/WayfireWM/wayfire/blob/2e0926f8f5a1c3834c51e97673bfbf074f4506ac/plugins/single_plugins/wsets.cpp#L25-L78

If you feel like it, you could also make this into a helper class in `plugins/common/...` - a node implementation which can take cairo options and text, can be set the geometry, etc. But if you feel like this is too much, feel free to also copy-paste the code and adapt it.

ammen99 reviewed (commented):
I have a few requests/suggestions on the implementation, but the general approach seems good!

ammen99 commented:
> * The swaylock indicator isn't displayed. Perhaps this is because swaylock uses a subsurface to draw it, and wlr-surface-node doesn't support child surfaces?

Yeah, you'd need to create a subsurface controller: https://github.com/WayfireWM/wayfire/blob/master/src/view/surface-impl.hpp#L15

Probably should be moved to the unstable APIs.

ammen99 commented:
> * Deal with outputs being resized

I think it might actually make sense to actually create two nodes: one node is compositor-generated, saying 'Waiting for screenlocker ...' and above it, we should add the screenlock.

This covers basically all cases: before the screenlock surface is there, we hide the contents of the desktop. If the screenlock crashes, the second node below will still cover / lock the screen. And, if we have to resize the output, the second node will cover the screen while the screenlocker is resizing.

lcolitti commented:
> > * The swaylock indicator isn't displayed. Perhaps this is because swaylock uses a subsurface to draw it, and wlr-surface-node doesn't support child surfaces?
> 
> Yeah, you'd need to create a subsurface controller: https://github.com/WayfireWM/wayfire/blob/master/src/view/surface-impl.hpp#L15
> 
> Probably should be moved to the unstable APIs.

Done in [this commit.](https://github.com/WayfireWM/wayfire/pull/2162/commits/b8028e620bbe6a7d7e21f8b72db356a8ffae7b84). I also renamed them, but I can keep the existing names if you prefer.

lcolitti commented:
> I think it might actually make sense to actually create two nodes: one node is compositor-generated, saying 'Waiting for screenlocker ...' and above it, we should add the screenlock.
> 
> This covers basically all cases: before the screenlock surface is there, we hide the contents of the desktop. If the screenlock crashes, the second node below will still cover / lock the screen. And, if we have to resize the output, the second node will cover the screen while the screenlocker is resizing.

I'm not sure if this is the best approach. For one thing, if we create the surface immediately, it will be visible/jarring. Ideally we just display the lockscreen surface like we do today without this plugin. Personally my lockscreen is the same as my background so when the lockscreen activates it just looks as if the windows smoothly fade out. I'd hate to make that less slick :smile:  I guess we could add this extra surface after a timeout? Not sure how to do that yet but there's probably a way.

Also, what would happen when an output is added? With the code as is, the plugin doesn't do anything. The screenlocker will probably react to this by creating a new surface, and when it does, the new output will be locked. But while the client creates the surface, I think there is a race where it's possible to display or interact with the new output, even if the others are locked. But can we close that race completely even if we create the surface in the compositor? There could still be a window between when the output is created and the surface is added.

ammen99 commented:
> But can we close that race completely even if we create the surface in the compositor? There could still be a window between when the output is created and the surface is added.

We most certainly can. The compositor is in full control of drawing every frame on every output. If you create a window/node on a new output in the added handler for that event, it will be present on the very first output frame :)

> I'd hate to make that less slick 😄 I guess we could add this extra surface after a timeout? Not sure how to do that yet but there's probably a way.

Ok, I guess this is a good point. Maybe the screenlocker wants to implement fade-in or a slide effect, so we shouldn't create the surface immediately. Maybe say 5 seconds after the lockscreen has kicked in? I'm not sure whether we want this to be configurable, so maybe just hardcode it and we can always add an option later.

lcolitti commented:
Ok, this looks closer to done now.

I think it deals correctly with outputs being added and removed, and (as required by the protocol) it now waits for the screenlocker to display a frame before it sends back the lock event. It uses @ammen99 's suggestion to use two nodes to ensure that if the screenlocker crashes, the screen remains locked. The UX of that is a bit primitive (a big, pixelated explosion emoji), but that shouldn't be too hard to fix by modifying simple-text-node.

It still also uses a surface instead of a view, but maybe that's OK for now.

ammen99 commented on plugins/protocols/session-lock.cpp:
```diff
@@ -0,0 +1,434 @@
+{
+    class wayfire_session_lock;
+
+    class lock_surface_node_t : public wf::scene::wlr_surface_node_t
```
Note that all nodes are shared_ptrs - this means that in general, it is possible that a node outlives the session-lock plugin to begin with, it can also outlive the output it is on, etc.

So, you should try to add code so that the object can 'work' (at least not crash) even as the underlying surface, lock, output, etc. are destroyed. I have tried to add some suggestions on how to accomplish this.

ammen99 commented on plugins/protocols/session-lock.cpp:
```diff
@@ -0,0 +1,434 @@
+                wf::wlr_surface_controller_t::try_free_controller(this->lock_surface->surface);
+                wf::scene::remove_child(shared_from_this());
+                lock_surface_destroy.disconnect();
+                this->lock->surface_destroyed(this->output);
```
Re lifetimes: I would personally set the destroy callback in the wayfire_session_lock class (this way, you don't have to pass the `lock` parameter.

ammen99 commented on plugins/protocols/session-lock.cpp:
```diff
@@ -0,0 +1,434 @@
+                lock_surface_destroy.disconnect();
+                this->lock->surface_destroyed(this->output);
+                const char *name = this->output->handle ? this->output->handle->name : "(deleted)";
+                LOGD("lock_surface on ", name, " destroyed");
```
You can/should reset the keyboard interaction to something else (something like this: https://github.com/WayfireWM/wayfire/blob/master/src/view/wlr-surface-node.cpp#L102) so that we don't get a crash if someone holds on to the node after the surface/output are destroyed.

ammen99 replied:
And also re lifetimes, you should set lock_surface to NULL here, and make sure it is not used in the other functions if it is NULL already (where it makes sense)

lcolitti replied:
Attempted to reset the keyboard interaction.

I don't think lock_surface can be used after it's destroyed. The only other method that uses it is attach_to_layer. But attach_to_layer will only be called if the output_state->surface is non-null. But when the lock_surface is destroyed, the callback for the destroy event calls this->lock->surface_destroyed, which nulls out the output_state->surface, so attach_to_layer will not be called.

ammen99 replied:
Ok, it might still be useful to set it to NULL so that if in the future we change anything, we directly get a null pointer crash instead of a dangling pointer.

ammen99 commented on plugins/protocols/session-lock.cpp:
```diff
@@ -0,0 +1,434 @@
+        void focus()
+        {
+            wf::get_core().seat->set_active_node(shared_from_this());
+            interaction->handle_keyboard_enter(wf::get_core().seat.get());
```
Why is this necessary? A simple `refocus()` call should be enough to focus the node. Manually calling `enter()` is usually not what you want, this is what core uses when it focuses your node.

ammen99 commented on plugins/protocols/session-lock.cpp:
```diff
@@ -0,0 +1,434 @@
+            void handle_keyboard_enter(wf::seat_t *seat)
+            {
+                wlr_seat_keyboard_enter(seat->seat, surface, nullptr, 0, nullptr);
+                LOGI("Focused ", output->name);
```
This should be a LOGD at most (though it probably should be removed altogether)

ammen99 commented on plugins/protocols/session-lock.cpp:
```diff
@@ -0,0 +1,434 @@
+        }
+
+        wf::keyboard_focus_node_t keyboard_refocus(wf::output_t *output) override
+        {
```
Shouldn't this focus only if our own output is focused?

ammen99 commented on plugins/protocols/session-lock.cpp:
```diff
@@ -0,0 +1,434 @@
+      private:
+        wayfire_session_lock *lock;
+        wlr_session_lock_surface_v1 *lock_surface;
+        wf::output_t *output;
```
Re lifetimes; I see that the output is a necessary thing to keep a reference to, which means you ought to also have a handler for when it is destroyed.

ammen99 commented on plugins/protocols/session-lock.cpp:
```diff
@@ -0,0 +1,434 @@
+            plugin(plugin), lock(lock)
+        {
+            auto& ol = wf::get_core().output_layout;
+            output_added.set_callback([this] (wf::output_added_signal *ev)
```
Instead of doing this manually, check out https://github.com/WayfireWM/wayfire/blob/master/src/api/wayfire/per-output-plugin.hpp#L30

You can add it as a base class and then it will connect to the signals, etc. for you.

ammen99 replied:
Looking at the code below, I assume you did this per hand simply because the mixin does not support non-plugin template types (yet). I think it should be very easy (and a good idea) to fix the mixin class - with an `if constexpr (std::is_base_v<per_output_plugin_instance_t, ConcretePluginType>) {}` around code blocks which are specific for per-output-plugin instances, so that the helper can be used generically.

lcolitti replied:
I did use the mixin for a while, to create the per-output objects. The reason I stopped using it is that the mixin expects the per-output instances to have zero-arg constructors. But here, the instances need to have a backpointer to a plugin-global object, for example to call notify_lock_state. Not sure if this is easy to do in the mixin - my command of C++ syntax isn't great.

ammen99 commented on plugins/protocols/session-lock.cpp:
```diff
@@ -0,0 +1,434 @@
+                wlr_session_lock_surface_v1 *lock_surface = (wlr_session_lock_surface_v1*)data;
+                wlr_output *wo = lock_surface->output;
+                wlr_session_lock_surface_v1_configure(lock_surface,
+                    wo->width / wo->scale,
```
I don't think this takes output transform into account. You can however find the `wf::output_t` associated with the wlr output and get its geometry - the geometry will be scaled and rotated correctly.

ammen99 commented on plugins/protocols/session-lock.cpp:
```diff
@@ -0,0 +1,434 @@
+                wlr_session_lock_surface_v1_configure(lock_surface,
+                    wo->width / wo->scale,
+                    wo->height / wo->scale);
+                LOGD("surface_configure output=", wo->name, " ", wo->width, "x", wo->height);
```
I think a lot of the log messages in this file can be put into a debugging category (see `debug.hpp`) - either VIEWS or LSHELL (I know this protocol technically is not layer-shell but they are closely related, at least in my mind :P)

ammen99 commented on plugins/protocols/session-lock.cpp:
```diff
@@ -0,0 +1,434 @@
+
+    void fini() override
+    {
+        // TODO: unlock everything?
```
I don't think we actually need to do anything. Plugins that cannot be unloaded like this one will be destroyed only at compositor shutdown. At that point the wayland display is being destroyed, so wlroots will auto-destroy the protocol and it will emit destruction signals for the various lock surfaces, etc. (I hope it does so at least).

ammen99 commented on plugins/protocols/session-lock.cpp:
```diff
@@ -0,0 +1,434 @@
+        void handle_output_removed(wf::output_t *output)
+        {
+            output_states.erase(output);
+            auto next_output_state = output_states[wf::get_core().seat->get_active_output()];
```
Again, is this needed? Core should auto-refocus correctly if you filter by the output in the refocus_keyboard() method.

ammen99 commented:
I went through your code, the overall design & implementation looks good. Lifetimes are hard to get right, especially when you have a shared_ptr, that needs some care.

lcolitti commented on plugins/common/wayfire/plugins/common/simple-text-node.hpp:
```diff
@@ -0,0 +1,71 @@
+
+    void set_text(std::string text)
+    {
+        cr_text.render_text(text, params);
```
Done, I think.

ammen99 commented:
@lcolitti Sorry I mistakenly merged this PR without even reviewing your latest changes .. Would you mind opening a new pull request or reopening this one? I'm not sure how to make GitHub do that ...

lcolitti commented:
Sorry I didn't update this PR for a while. I've been travelling and can't test with my multi-monitor setup.

I don't think there's a way to reopen this after the merge. Opened #2237 instead. @ammen99 can you take a look? The code there addresses most of your review comments, including hopefully not crashing on shutdown. I've been running this code on my machine for a while and it seems to work well.
